server/CGameHandler.cpp

```diff
@@ -2233,6 +2242,53 @@ bool CGameHandler::razeStructure (ObjectInstanceID tid, BuildingID bid)
 	return true;
 }
 
+bool CGameHandler::spellResearch(ObjectInstanceID tid, SpellID spellAtSlot)
+{
+	if(!getSettings().getBoolean(EGameSettings::TOWNS_SPELL_RESEARCH) && complain("Spell research not allowed!"))
+		return false;
+
+	CGTownInstance *t = gs->getTown(tid);
+
+	int level = -1;
+	for(int i = 0; i < t->spells.size(); i++)
+		if(vstd::find_pos(t->spells[i], spellAtSlot) != -1)
+			level = i;
+	
+	if(level == -1 && complain("Spell for replacement not found!"))
+		return false;
+	
+	int daysSinceLastResearch = gs->getDate(Date::DAY) - t->lastSpellResearchDay;
+	if(!daysSinceLastResearch && complain("Already researched today!"))
+		return false;
+
+	TResources cost;
+	cost[EGameResID::GOLD] = 1000;
+	cost[EGameResID::MERCURY] = (level + 1) * 2;
+	cost[EGameResID::SULFUR] = (level + 1) * 2;
+	cost[EGameResID::CRYSTAL] = (level + 1) * 2;
+	cost[EGameResID::GEMS] = (level + 1) * 2;
```

Comment:
Should be in GameSettings. For example as two values: base cost and cost / level. 
I think ResourceSet already has constructor from json